core/server/adapters/scheduling/post-scheduling/index.js

```diff
@@ -1,24 +1,79 @@
-const Promise = require('bluebird'),
-    moment = require('moment'),
-    localUtils = require('../utils'),
-    common = require('../../../lib/common'),
-    models = require('../../../models'),
-    urlUtils = require('../../../lib/url-utils'),
-    _private = {};
+const Promise = require('bluebird');
+const moment = require('moment');
+const jwt = require('jsonwebtoken');
+const localUtils = require('../utils');
+const common = require('../../../lib/common');
+const models = require('../../../models');
+const urlUtils = require('../../../lib/url-utils');
+const _private = {};
+const SCHEDULED_RESOURCES = ['post', 'page'];
+
+/**
+ * @description Load the internal scheduler integration
+ *
+ * @return {Promise}
+ */
+_private.getSchedulerIntegration = function () {
+    return models.Integration.findOne({slug: 'ghost-scheduler'}, {withRelated: 'api_keys'})
+        .then((integration) => {
+            if (!integration) {
+                throw new common.errors.NotFoundError({
+                    message: common.i18n.t('errors.api.resource.resourceNotFound', {
+                        resource: 'Integration'
+                    })
+                });
+            }
+            return integration.toJSON();
+        });
+};
+
+/**
+ * @description Get signed admin token for making authenticated scheduling requests
+ *
+ * @return {Promise}
+ */
+_private.getSignedAdminToken = function (options) {
+    const {model, apiUrl, integration} = options;
+    let key = integration.api_keys[0];
+
+    const JWT_OPTIONS = {
+        keyid: key.id,
+        algorithm: 'HS256',
+        audience: apiUrl
+    };
+
+    // Default token expiry is till 10 minutes after scheduled time
+    // or if published_at is in past then till 10 minutes after blog start
+    // and never before 10 mins to publish time
+    let tokenExpiry = moment(model.get('published_at')).add(10, 'm');
+    if (tokenExpiry.isBefore(moment())) {
+        tokenExpiry = moment().add(10, 'm');
+    }
```

Comment:
There is not much logic behind the window being 10 minutes either side of boot/publish time than this seemed like a reasonable enough window for schedules to run. 🤷‍♂

Reply:
I'd suggest we make it longer because we want to cover retries which could be caused by networking problems and also the scheduler service going down for any reason.

Maybe a window of 10min before and 6hr after?

If the scheduler service fails to make the request then the post will stay unpublished until Ghost is restarted.

Reply:
> Maybe a window of 10min before and 6hr after?
Hmm yeah, was not sure how much window is good enough for retries, but agree having it on higher side is definitely better failsafe in this case as the next restart might be after much longer duration. I'll update the duration 👍

Reply:
@kevinansfield Updated